service/matching/membership.go

```diff
@@ -0,0 +1,91 @@
+package matching
+
+import (
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/membership"
+	"github.com/uber/cadence/common/service"
+	"github.com/uber/cadence/service/matching/handler"
+
+	"github.com/uber/cadence/service/matching/config"
+)
+
+func (s *Service) subscribeToMembershipChanges(
+	engine handler.Engine,
+) {
+	listener := make(chan *membership.ChangedEvent)
+	s.GetMembershipResolver().Subscribe(service.Matching, "matching-engine", listener)
+	self, err := s.GetMembershipResolver().WhoAmI()
+	if err != nil {
+		s.GetLogger().Error("Failed to get self from hashring")
+		return
+	}
+	for {
+		select {
+		case event := <-listener:
+			checkAndShutdownIfEvicted(engine, s.config, s.GetLogger(), self, event)
+		case <-s.stopC:
+			return
+		}
+	}
+}
+
+// checkAndShutdownIfEvicted checks if the host has been evicted from the hashring
+func checkAndShutdownIfEvicted(engine handler.Engine, cfg *config.Config, log log.Logger, self membership.HostInfo, event *membership.ChangedEvent) {
+	featureEnabled := cfg.EnableServiceDiscoveryShutdown()
+	if selfIsEvicted(log, self, event) {
+		if featureEnabled {
+			log.Info("shutting down service due to hashring removing it from service", tag.MembershipChangeEvent(event))
+			engine.Stop()
```

Comment:
Matching Handler will not know whether engine is stopped or not and it will still call Add/Poll methods on engine. There can be some edge cases and undefined behaviors of doing so. I'd consider stopping the handler itself and also interfering with the calls (either in a middleware or in handler) to reject them by returning "AlreadyStopped" error. 

Let's also add an integration test for this

Reply:
ah, I was going to make engine return errors, but I can stop the handler as well, that's a fair point, I think the errors would bubble up in either case, but you're right about the behaviour of being partially shut down being a bit undefined.

Reply:
hm.. actually that might also crashloop i think, because that'll fail healthchecks